contracts/meta-staking/src/contract.rs

```diff
@@ -174,32 +197,127 @@ mod execute {
         Ok(Response::default().add_message(msg))
     }
 
-    // TODO finish me
+    enum Method {
+        Add,
+        Sub,
+    }
+
+    fn update_delegations<'a>(
+        deps: DepsMut<'a>,
+        info: MessageInfo,
+        validator: &str,
+        amount: Uint128,
+        method: Method,
+    ) -> Result<DepsMut<'a>, ContractError> {
+        let action = |validator_info: Option<Uint128>| -> Result<_, ContractError> {
+            match method {
+                Method::Sub => {
+                    let val = validator_info.ok_or(ContractError::NoDelegationsForValidator {})?;
+                    val.checked_sub(amount)
+                        .map_err(|_| ContractError::InsufficientDelegation {})
+                }
+                Method::Add => Ok(validator_info.unwrap_or_default() + amount),
+            }
+        };
+
+        VALIDATORS_BY_CONSUMER.update(deps.storage, (&info.sender, validator), action)?;
+
+        CONSUMERS_BY_VALIDATOR.update(deps.storage, (validator, &info.sender), action)?;
+
+        Ok(deps)
+    }
+
     pub fn withdraw_delegator_reward(
+        deps: DepsMut,
```

Comment:
Question:

This is a useful function and I like what it does. But what case do we just want to withdraw reward here and not send to the consumers?

I could see this as an internal function for accounting, and maybe we do withdraw more often. But if an external user calls something, it probably should be "withdraw from validator and send to all consumers" (`withdraw_to_customer`)

Reply:
My short suggestion (which can be a future refactor) would be:

1. Make `withdraw_delegator_reward` an internal function (not exposed to message, but maybe called upon increased staking for example).
2. `withdraw_to_customer` should call this `withdraw_delegator_reward` function first, then send all rewards to the consumer. That is the only one exposed to the end-users, rather than give them a 2 step process.

Note: this is just a suggestion on architecture refinement, what you have works. I won't block this PR on it.

Reply:
I totally agree with this design, now after understanding more the whole process this seems to be the right one.
This is very confusing because we distribute rewards on Chain A, but the rewards are on Chain B, and I actually forgot a step that make this whole process not an accurate reward distribution.

The whole process should be like this started by 1 call:
1. we do stake/unstake on provider and send a IBC call to consumer.
2. Then on consumer we start the `withdraw_to_customer` of meta-staking to get the most recent rewards based on the old stake.
3. `withdraw_to_customer` starts the whole process to send rewards to provider.
4.  on provider when we receive the rewards, we then do the reward calculation of the initial delegator who started this process and of the validator.

@ethanfrey I'm pretty sure its confusing, but basically right now, on provider we are lagging behind the rewards from meta-staking, so the provider calculations are always with a different amount of stake and rewards then what actually happen on meta-staking.

To fix that, we need to do a IBC call each time stake/unstake and withdraw rewards happen on provider, get the latest rewards from meta-staking, and only then do the provider calculations when rewards received. 

Not sure if we want this fix in this PR or not, but it is something we need to make this accurate.

Reply:
Also with the above fix we don't need to expose `withdraw_delegator_reward`, and do it like you suggested, make it as an internal function that is called inside `withdraw_to_customer`.